libvast/src/option_declaration_set.cpp

```diff
@@ -0,0 +1,296 @@
+
+option_declaration_set::option_declaration::option_declaration(
+  std::string_view long_name, std::vector<char> short_names,
+  std::string_view description, bool has_argument, data default_value)
+  : long_name_(long_name),
+    short_names_(std::move(short_names)),
+    description_(description),
+    has_argument_(has_argument),
+    default_value_(std::move(default_value)) {
+  // nop
+}
+
+std::pair<option_declaration_set::parse_state, data>
+option_declaration_set::option_declaration::parse(
+  std::string_view value) const {
+  auto result = visit(detail::overload(
+    [&](const auto& arg) {
+      using arg_type = std::decay_t<decltype(arg)>;
+      auto x = to<arg_type>(value);
+      if (!x)
+        // FIXME: We lose valuable error informaton here.
+        return std::make_pair(parse_state::failed_to_parse_argument,
+                              default_value());
+      return std::make_pair(parse_state::successful, data{*x});
+    },
+    // TODO: These overloads are nessesary as no respective parser exists at the 
+    // moment. Remove me when possible.
+    [&](const none&) {
+      return std::make_pair(parse_state::type_not_parsebale,
+                              default_value());
+    },
+    [&](const set&) {
+      return std::make_pair(parse_state::type_not_parsebale,
+                              default_value());
+    },
+    [&](const table&) {
+      return std::make_pair(parse_state::type_not_parsebale,
+                              default_value());
+    },
+    [&](const vector&) {
+      return std::make_pair(parse_state::type_not_parsebale,
+                              default_value());
+    }), default_value_);
+  return result; 
+}
+
+const std::string&
+option_declaration_set::option_declaration::long_name() const {
+  return long_name_;
+}
+
+const std::vector<char>&
+option_declaration_set::option_declaration::short_names() const {
+  return short_names_;
+}
+
+const std::string&
+option_declaration_set::option_declaration::description() const {
+  return description_;
+}
+
+bool option_declaration_set::option_declaration::has_argument() const {
+  return has_argument_;
+}
+
+const data& option_declaration_set::option_declaration::default_value() const {
+  return default_value_;
+}
+
+option_declaration_set::option_declaration_set() {
+  add("help,h?", "print this text", false);
+}
+
+optional<const option_declaration_set::option_declaration&>
+option_declaration_set::find(std::string_view name) const {
+  // TODO: Remove explicit conversion to a string
+  if (auto it = long_opts_.find(std::string{name}); it != long_opts_.end())
+    return *it->second;
+  return {};
+}
+
+std::string option_declaration_set::usage() const {
+  //<--- argument ---> <---- desciption ---->
+  //-w [--write] arg  : path to write events to
+  auto build_argument = [](const option_declaration& x) {
+    std::stringstream arg;
+    auto& shorts = x.short_names();
+    arg << "  "; 
+    if (!shorts.empty()) {
```

Comment:
Instead of first checking for not empty and then checking `size > 1`, I would rewrite it like this:

```cpp
if (!shorts.empty()) {
  auto i = shorts.begin();
  auto e = shorts.end();
  arg << "-" << *i << " [";
  for (++i; i != e; ++i)
    arg << "-" << *i << ", ";
  arg << '--' << x.long_name() << ']';
} else {
  arg << '--' << x.long_name();
}
```